toolchains/tool.bzl

```diff
@@ -0,0 +1,115 @@
+# Copyright 2023 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Implementation of tool"""
+
+load("//toolchains/private:collect.bzl", "collect_data")
+load(
+    ":toolchain_info.bzl",
+    "ToolCapabilityInfo",
+    "ToolInfo",
+)
+
+def _tool_impl(ctx):
+    exe_info = ctx.attr.src[DefaultInfo]
+    if exe_info.files_to_run != None and exe_info.files_to_run.executable != None:
+        exe = exe_info.files_to_run.executable
+    elif len(exe_info.files.to_list()) == 1:
+        exe = exe_info.files.to_list()[0]
+    else:
+        fail("Expected tool's src attribute to be either an executable or a single file")
+
+    data = collect_data(ctx.attr.data + [ctx.attr.src])
+    tool = ToolInfo(
+        label = ctx.label,
+        exe = exe,
+        files_to_run = data,
+        execution_requirements = ctx.attr.execution_requirements,
+        capabilities = tuple([target[ToolCapabilityInfo] for target in ctx.attr.capabilities]),
+    )
+
+    link = ctx.actions.declare_file(ctx.label.name)
+    ctx.actions.symlink(
+        output = link,
+        target_file = exe,
+        is_executable = True,
+    )
+    return [
+        tool,
+        # This isn't required, but now we can do "bazel run <tool>", which can
+        # be very helpful when debugging toolchains.
+        DefaultInfo(
+            files = depset([link]),
+            runfiles = ctx.runfiles(transitive_files = depset(transitive = [element for element in data if type(element) == "depset"])),
+            executable = link,
+        ),
+    ]
+
+tool = rule(
+    implementation = _tool_impl,
+    # @unsorted-dict-items
+    attrs = {
+        "src": attr.label(
+            allow_files = True,
+            doc = """The underlying binary that this tool represents.
+
+Usually just a single prebuilt (eg. @toolchain//:bin/clang), but may be any
+executable label.
+""",
+        ),
+        "data": attr.label_list(
+            allow_files = True,
+            doc = """Additional files that are required for this tool to run.
```

Comment:
As mentioned in my comment on `collect.bzl`, this doesn't behave like a typical `data` attribute. It's also not clear how a single executable in `src` should find anything listed in `data` - by relative path or using standard runfiles lookups? I am a bit worried that we are codifying C++ toolchain weirdness here.

Reply:
Actually, this has nothing to do with C++ toolchains. My intent was for it to behave like a typical data attribute, I just wasn't particularly familiar with how to do it.